lib/pwned_password_downloader.rb

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env ruby
+require 'fileutils'
+require 'net/http/persistent'
+require 'ruby-progressbar'
+
+class PwnedPasswordDownload
+  attr_reader :destination,
+              :num_threads,
+              :keep_threshold
+
+  def initialize(
+    destination: 'tmp/pwned',
+    num_threads: 64,
+    keep_threshold: 30
+  )
+    @destination = destination
+    @num_threads = num_threads
+    @keep_threshold = keep_threshold
+  end
+
+  def run!
+    queue = Queue.new
+
+    ('00000'.to_i(16)..'FFFFF'.to_i(16)).each do |prefix_num|
+      queue << prefix_num.to_s(16).upcase.rjust(5, '0')
+    end
+
+    bar = ProgressBar.create(
+      title: 'Downloading...',
+      total: queue.size,
+      format: '[ %t ] %p%% %B %a (%E)',
+    )
+
+    FileUtils.mkdir_p(destination)
+
+    num_threads.times.map do
+      Thread.new do |thread_id|
+        net_http = Net::HTTP::Persistent.new(name: "thread_id_#{thread_id}")
+
+        while (prefix = queue.pop)
+          bar.increment
+          next if already_downloaded?(prefix)
+          write_one(prefix:, content: download_one(prefix:, net_http:))
+        end
+      ensure
+        net_http.shutdown
+      end
+    end
+
+    sleep 3 until queue.empty?
```

Comment:
note for testing, we'll def want to do
```
allow(downloader).to receive(:sleep)
```
so it doesn't actually sleep

Reply:
> note for testing, we'll def want to do

This was included in 371df2768d8fb51241496593e04e197e435fc186.